File: chemometrics/modeling.py
"""
Módulo de modelagem quimiométrica.
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.cross_decomposition import PLSRegression
from sklearn.metrics import r2_score, mean_squared_error
from sklearn.model_selection import cross_val_score
from typing import Dict, List, Optional, Tuple
import logging

logger = logging.getLogger(__name__)


class PLSRegressor:
    """
    Regressor PLS com funcionalidades para análise quimiométrica.
    """
    
    def __init__(self, n_components: Optional[int] = None, max_iter: int = 500,
                 tol: float = 1e-06, copy: bool = True, scale: bool = True):
        """
        Inicializa o regressor PLS.
        
        Parameters
        ----------
        n_components : int, optional
            Número de componentes latentes. Se None, será determinado por validação cruzada.
        max_iter : int, default=500
            Número máximo de iterações.
        tol : float, default=1e-06
            Tolerância para convergência.
        copy : bool, default=True
            Se True, faz uma cópia dos dados.
        scale : bool, default=True
            Se True, escala os dados.
        """
        self.n_components = n_components
        self.max_iter = max_iter
        self.tol = tol
        self.copy = copy
        self.scale = scale
        
        self.model_ = None
        self.is_fitted = False
        self.metrics_ = {}
        self.vip_scores_ = None
        
    def fit(self, X: np.ndarray, y: np.ndarray, 
            cv_folds: int = 5, optimize_components: bool = True) -> 'PLSRegressor':
        """
        Ajusta o modelo PLS aos dados.
        
        Parameters
        ----------
        X : np.ndarray
            Dados de entrada.
        y : np.ndarray
            Variável resposta.
        cv_folds : int, default=5
            Número de folds para validação cruzada.
        optimize_components : bool, default=True
            Se True, otimiza o número de componentes por validação cruzada.
            
        Returns
        -------
        self : PLSRegressor
            Instância ajustada.
        """
        X, y = np.array(X), np.array(y)
        
        if X.ndim == 1:
            X = X.reshape(-1, 1)
        if y.ndim == 1:
            y = y.reshape(-1, 1)
            
        # Otimizar número de componentes se necessário
        if optimize_components and self.n_components is None:
            self.n_components = self._optimize_components(X, y, cv_folds)
            
        # Ajustar modelo final
        self.model_ = PLSRegression(
            n_components=self.n_components,
            max_iter=self.max_iter,
            tol=self.tol,
            copy=self.copy,
            scale=self.scale
        )
        
        self.model_.fit(X, y)
        self.is_fitted = True
        
        # Calcular VIP scores
        self.vip_scores_ = self._calculate_vip_scores(X)
        
        logger.info(f"Modelo PLS ajustado com {self.n_components} componentes")
        
        return self
    
    def predict(self, X: np.ndarray) -> np.ndarray:
        """Faz predições usando o modelo ajustado."""
        if not self.is_fitted:
            raise ValueError("Modelo não foi ajustado. Chame fit() primeiro.")
            
        return self.model_.predict(X)
    
    def score(self, X: np.ndarray, y: np.ndarray) -> float:
        """Calcula o R² score."""
        if not self.is_fitted:
            raise ValueError("Modelo não foi ajustado. Chame fit() primeiro.")
            
        return self.model_.score(X, y)
    
    def evaluate(self, X_cal: np.ndarray, y_cal: np.ndarray,
                X_test: np.ndarray, y_test: np.ndarray,
                unit: str = 'units') -> Dict[str, float]:
        """
        Avalia o modelo com métricas completas.
        
        Parameters
        ----------
        X_cal : np.ndarray
            Dados de calibração.
        y_cal : np.ndarray
            Resposta de calibração.
        X_test : np.ndarray
            Dados de teste.
        y_test : np.ndarray
            Resposta de teste.
        unit : str, default='units'
            Unidade de medida.
            
        Returns
        -------
        metrics : dict
            Métricas de avaliação.
        """
        if not self.is_fitted:
            raise ValueError("Modelo não foi ajustado. Chame fit() primeiro.")
            
        # Predições
        y_cal_pred = self.predict(X_cal)
        y_test_pred = self.predict(X_test)
        
        # Métricas de calibração
        r2c = r2_score(y_cal, y_cal_pred)
        rmsec = np.sqrt(mean_squared_error(y_cal, y_cal_pred))
        
        # Métricas de predição
        r2p = r2_score(y_test, y_test_pred)
        rmsep = np.sqrt(mean_squared_error(y_test, y_test_pred))
        
        # Validação cruzada
        cv_scores = cross_val_score(self.model_, X_cal, y_cal, cv=5, scoring='r2')
        rmsecv = np.sqrt(mean_squared_error(y_cal, self.predict(X_cal)))
        
        self.metrics_ = {
            'R2c': r2c,
            'R2p': r2p,
            'RMSEC': rmsec,
            'RMSEP': rmsep,
            'RMSECV': rmsecv,
            'n_components': self.n_components,
            'unit': unit
        }
        
        return self.metrics_
    
    def plot_results(self, X_cal: np.ndarray, y_cal: np.ndarray,
                    X_test: np.ndarray, y_test: np.ndarray,
                    unit: str = 'units', figsize: Tuple[int, int] = (15, 6)) -> None:
        """
        Plota os resultados do modelo.
        
        Parameters
        ----------
        X_cal : np.ndarray
            Dados de calibração.
        y_cal : np.ndarray
            Resposta de calibração.
        X_test : np.ndarray
            Dados de teste.
        y_test : np.ndarray
            Resposta de teste.
        unit : str, default='units'
            Unidade de medida.
        figsize : tuple, default=(15, 6)
            Tamanho da figura.
        """
        if not self.is_fitted:
            raise ValueError("Modelo não foi ajustado. Chame fit() primeiro.")
            
        # Predições
        y_cal_pred = self.predict(X_cal)
        y_test_pred = self.predict(X_test)
        
        # Métricas
        r2c = r2_score(y_cal, y_cal_pred)
        r2p = r2_score(y_test, y_test_pred)
        rmsec = np.sqrt(mean_squared_error(y_cal, y_cal_pred))
        rmsep = np.sqrt(mean_squared_error(y_test, y_test_pred))
        
        # Criar figura
        fig, (ax1, ax2) = plt.subplots(1, 2, figsize=figsize)
        
        # Gráfico Predito vs Real
        ax1.scatter(y_cal, y_cal_pred, alpha=0.7, label='Calibração', color='blue', s=50)
        ax1.scatter(y_test, y_test_pred, alpha=0.7, label='Teste', color='red', s=50)
        
        # Linha de referência
        min_val = min(np.min(y_cal), np.min(y_test), np.min(y_cal_pred), np.min(y_test_pred))
        max_val = max(np.max(y_cal), np.max(y_test), np.max(y_cal_pred), np.max(y_test_pred))
        ax1.plot([min_val, max_val], [min_val, max_val], 'k--', alpha=0.8, linewidth=2)
        
        ax1.set_xlabel(f'Real ({unit})')
        ax1.set_ylabel(f'Predito ({unit})')
        ax1.legend()
        ax1.set_title(f'Predito vs Real\nR²c = {r2c:.3f}, R²p = {r2p:.3f}')
        ax1.grid(True, alpha=0.3)
        
        # Gráfico de Resíduos
        residuals_cal = y_cal - y_cal_pred
        residuals_test = y_test - y_test_pred
        
        ax2.scatter(y_cal_pred, residuals_cal, alpha=0.7, label='Calibração', color='blue', s=50)
        ax2.scatter(y_test_pred, residuals_test, alpha=0.7, label='Teste', color='red', s=50)
        ax2.axhline(y=0, color='k', linestyle='--', alpha=0.8, linewidth=2)
        
        ax2.set_xlabel(f'Predito ({unit})')
        ax2.set_ylabel(f'Resíduos ({unit})')
        ax2.legend()
        ax2.set_title(f'Resíduos\nRMSEC = {rmsec:.3f}, RMSEP = {rmsep:.3f}')
        ax2.grid(True, alpha=0.3)
        
        plt.tight_layout()
        plt.show()
    
    def plot_vip_scores(self, feature_names: Optional[List[str]] = None,
                       threshold: float = 1.0, figsize: Tuple[int, int] = (12, 6)) -> None:
        """
        Plota os VIP scores.
        
        Parameters
        ----------
        feature_names : list, optional
            Nomes das variáveis.
        threshold : float, default=1.0
            Limiar para destacar variáveis importantes.
        figsize : tuple, default=(12, 6)
            Tamanho da figura.
        """
        if self.vip_scores_ is None:
            raise ValueError("VIP scores não foram calculados. Ajuste o modelo primeiro.")
            
        plt.figure(figsize=figsize)
        
        if feature_names is None:
            feature_names = [f'Var_{i}' for i in range(len(self.vip_scores_))]
        
        # Plotar VIP scores
        bars = plt.bar(range(len(self.vip_scores_)), self.vip_scores_, 
                      color=['red' if score > threshold else 'blue' for score in self.vip_scores_])
        
        # Linha de referência
        plt.axhline(y=threshold, color='red', linestyle='--', alpha=0.7, 
                   label=f'Limiar = {threshold}')
        
        plt.xlabel('Variáveis')
        plt.ylabel('VIP Score')
        plt.title('VIP Scores - Importância das Variáveis')
        plt.legend()
        plt.grid(True, alpha=0.3)
        
        # Rotacionar labels se necessário
        if len(feature_names) > 20:
            plt.xticks(rotation=45)
        
        plt.tight_layout()
        plt.show()
    
    def get_results_dataframe(self, X_cal: np.ndarray, y_cal: np.ndarray,
                            X_test: np.ndarray, y_test: np.ndarray) -> pd.DataFrame:
        """
        Cria DataFrame com resultados detalhados.
        
        Parameters
        ----------
        X_cal : np.ndarray
            Dados de calibração.
        y_cal : np.ndarray
            Resposta de calibração.
        X_test : np.ndarray
            Dados de teste.
        y_test : np.ndarray
            Resposta de teste.
            
        Returns
        -------
        df_results : pd.DataFrame
            DataFrame com resultados.
        """
        if not self.is_fitted:
            raise ValueError("Modelo não foi ajustado. Chame fit() primeiro.")
            
        # Predições
        y_cal_pred = self.predict(X_cal)
        y_test_pred = self.predict(X_test)
        
        # Criar DataFrame
        df_results = pd.DataFrame({
            'conjunto': ['calibração'] * len(y_cal) + ['teste'] * len(y_test),
            'y_real': np.concatenate([y_cal, y_test]),
            'y_predito': np.concatenate([y_cal_pred, y_test_pred]),
            'erro_absoluto': np.abs(np.concatenate([
                y_cal - y_cal_pred, y_test - y_test_pred
            ])),
            'erro_percentual': np.abs(np.concatenate([
                np.where(y_cal != 0, (y_cal - y_cal_pred) / y_cal * 100, 0),
                np.where(y_test != 0, (y_test - y_test_pred) / y_test * 100, 0)
            ]))
        })
        
        return df_results
    
    def _optimize_components(self, X: np.ndarray, y: np.ndarray, 
                           cv_folds: int = 5) -> int:
        """Otimiza o número de componentes por validação cruzada."""
        n_comp_max = min(10, X.shape[0] - 1, X.shape[1])
        best_score = -np.inf
        best_n_comp = 1
        
        for n_comp in range(1, n_comp_max + 1):
            try:
                pls_temp = PLSRegression(n_components=n_comp)
                scores = cross_val_score(pls_temp, X, y, cv=cv_folds, scoring='r2')
                mean_score = np.mean(scores)
                
                if mean_score > best_score:
                    best_score = mean_score
                    best_n_comp = n_comp
                    
            except Exception as e:
                logger.warning(f"Erro ao testar {n_comp} componentes: {e}")
                continue
                
        logger.info(f"Número ótimo de componentes: {best_n_comp} (R² = {best_score:.3f})")
        return best_n_comp
    
    def _calculate_vip_scores(self, X: np.ndarray) -> np.ndarray:
        """Calcula VIP scores para seleção de variáveis."""
        if not self.is_fitted:
            return None
            
        t = self.model_.x_scores_
        w = self.model_.x_weights_
        q = self.model_.y_loadings_
        
        p = X.shape[1]
        vip = np.zeros(p)
        
        s = np.diag(t.T @ t @ q.T @ q).reshape(-1, 1)
        total_s = np.sum(s)
        
        for i in range(p):
            weight = np.array([(w[i, j] / np.linalg.norm(w[:, j]))**2 
                             for j in range(self.model_.n_components)])
            vip[i] = np.sqrt(p * (s.T @ weight) / total_s)
        
        return vip.ravel()

File: MATLAB/chemometrics/modeling.m
function model = modeling(X, y, method, varargin)
%MODELING Treina modelos de regressão quimiométrica
%
% Sintaxe:
%   model = modeling(X, y, method)
%   model = modeling(X, y, method, 'Name__', Value)
%
% Parâmetros:
%   X - Dados de entrada (amostras x variáveis)
%   y - Variável resposta (amostras x 1)
%   method - Método de modelagem:
%           'pls' - Partial Least Squares
%
% Propriedades (Name-Value pairs):
%   'n_components' - número de componentes (default: auto)
%   'cv_folds' - número de folds para validação cruzada (default: 5)
%   'optimize_components' - otimizar componentes automaticamente (default: true)
%
% Exemplo:
%   model = modeling(X, y, 'pls', 'n_components', 5);

    % Configurar parser de argumentos
    p = inputParser;
    addRequired(p, 'X', @isnumeric);
    addRequired(p, 'y', @isnumeric);
    addRequired(p, 'method', @ischar);
    addParameter(p, 'n_components', [], @isnumeric);
    addParameter(p, 'cv_folds', 5, @isnumeric);
    addParameter(p, 'optimize_components', true, @islogical);
    
    parse(p, X, y, method, varargin{:});
    
    % Validar entrada
    if isempty(X) || isempty(y)
        error('Dados de entrada não podem estar vazios.');
    end
    
    if size(X, 1) ~= size(y, 1)
        error('Número de amostras em X e y deve ser igual.');
    end
    
    % Aplicar método de modelagem
    switch lower(method)
        case 'pls'
            model = train_pls(X, y, p.Results.n_components, ...
                            p.Results.cv_folds, p.Results.optimize_components);
        otherwise
            error('Método de modelagem não reconhecido: %s', method);
    end
end

function pls_model = train_pls(X, y, n_components, cv_folds, optimize_components)
% Treina modelo PLS
    % Otimizar número de componentes se necessário
    if optimize_components && isempty(n_components)
        n_components = optimize_pls_components(X, y, cv_folds);
    elseif isempty(n_components)
        n_components = min(10, size(X, 1) - 1, size(X, 2));
    end
    
    % Treinar modelo PLS
    try
        [XL, YL, XS, YS, BETA, PCTVAR, MSE, stats] = plsregress(X, y, n_components);
        
        % Criar estrutura do modelo
        pls_model = struct();
        pls_model.method = 'pls';
        pls_model.n_components = n_components;
        pls_model.XL = XL;
        pls_model.YL = YL;
        pls_model.XS = XS;
        pls_model.YS = YS;
        pls_model.BETA = BETA;
        pls_model.PCTVAR = PCTVAR;
        pls_model.MSE = MSE;
        pls_model.stats = stats;
        pls_model.is_fitted = true;
        
        % Calcular VIP scores
        pls_model.vip_scores = calculate_vip_scores(X, XL, YL);
        
        fprintf('Modelo PLS treinado com %d componentes\n', n_components);
        
    catch ME
        error('Erro ao treinar modelo PLS: %s', ME.message);
    end
end

function n_comp_opt = optimize_pls_components(X, y, cv_folds)
% Otimiza o número de componentes por validação cruzada
    n_comp_max = min(10, size(X, 1) - 1, size(X, 2));
    best_score = -inf;
    best_n_comp = 1;
    
    fprintf('Otimizando número de componentes...\n');
    
    for n_comp = 1:n_comp_max
        try
            % Validação cruzada
            cv_scores = zeros(cv_folds, 1);
            indices = crossvalind('Kfold', size(X, 1), cv_folds);
            
            for fold = 1:cv_folds
                test_idx = (indices == fold);
                train_idx = ~test_idx;
                
                % Treinar modelo
                [XL, YL, XS, YS, BETA] = plsregress(X(train_idx, :), y(train_idx), n_comp);
                
                % Predizer
                y_pred = [ones(sum(test_idx), 1), X(test_idx, :)] * BETA;
                
                % Calcular R²
                cv_scores(fold) = calculate_r2(y(test_idx), y_pred);
            end
            
            mean_score = mean(cv_scores);
            
            if mean_score > best_score
                best_score = mean_score;
                best_n_comp = n_comp;
            end
            
        catch
            continue;
        end
    end
    
    fprintf('Número ótimo de componentes: %d (R² = %.3f)\n', best_n_comp, best_score);
    n_comp_opt = best_n_comp;
end

function vip = calculate_vip_scores(X, XL, YL)
% Calcula VIP scores para seleção de variáveis
    p = size(X, 2);
    n_comp = size(XL, 2);
    
    vip = zeros(p, 1);
    
    % Calcular VIP scores
    for i = 1:p
        weight_sum = 0;
        total_sum = 0;
        
        for j = 1:n_comp
            weight = (XL(i, j) / norm(XL(:, j)))^2;
            weight_sum = weight_sum + weight * YL(j)^2;
            total_sum = total_sum + YL(j)^2;
        end
        
        vip(i) = sqrt(p * weight_sum / total_sum);
    end
end

function y_pred = predict(model, X)
% Faz predições usando o modelo ajustado
    if ~model.is_fitted
        error('Modelo não foi ajustado. Chame modeling() primeiro.');
    end
    
    switch model.method
        case 'pls'
            y_pred = [ones(size(X, 1), 1), X] * model.BETA;
        otherwise
            error('Método de modelo não reconhecido: %s', model.method);
    end
end

function metrics = evaluate_model(model, X_cal, y_cal, X_test, y_test, unit)
% Avalia o modelo com métricas completas
    if nargin < 6
        unit = 'units';
    end
    
    if ~model.is_fitted
        error('Modelo não foi ajustado. Chame modeling() primeiro.');
    end
    
    % Predições
    y_cal_pred = predict(model, X_cal);
    y_test_pred = predict(model, X_test);
    
    % Métricas de calibração
    r2c = calculate_r2(y_cal, y_cal_pred);
    rmsec = calculate_rmse(y_cal, y_cal_pred);
    
    % Métricas de predição
    r2p = calculate_r2(y_test, y_test_pred);
    rmsep = calculate_rmse(y_test, y_test_pred);
    
    % Criar estrutura de métricas
    metrics = struct();
    metrics.R2c = r2c;
    metrics.R2p = r2p;
    metrics.RMSEC = rmsec;
    metrics.RMSEP = rmsep;
    metrics.n_components = model.n_components;
    metrics.unit = unit;
    
    % Exibir métricas
    print_metrics(metrics, 'Métricas do Modelo PLS');
end

function r2 = calculate_r2(y_true, y_pred)
% Calcula coeficiente de determinação R²
    ss_res = sum((y_true - y_pred).^2);
    ss_tot = sum((y_true - mean(y_true)).^2);
    r2 = 1 - (ss_res / ss_tot);
end

function rmse = calculate_rmse(y_true, y_pred)
% Calcula Root Mean Square Error
    rmse = sqrt(mean((y_true - y_pred).^2));
end

function print_metrics(metrics, title)
% Imprime as métricas de forma formatada
    fprintf('\n=== %s ===\n', upper(title));
    fprintf('R²c: %.4f\n', metrics.R2c);
    fprintf('R²p: %.4f\n', metrics.R2p);
    fprintf('RMSEC: %.4f %s\n', metrics.RMSEC, metrics.unit);
    fprintf('RMSEP: %.4f %s\n', metrics.RMSEP, metrics.unit);
    fprintf('Componentes: %d\n', metrics.n_components);
end

function plot_results(model, X_cal, y_cal, X_test, y_test, unit)
% Plota os resultados do modelo
    if nargin < 6
        unit = 'units';
    end
    
    % Predições
    y_cal_pred = predict(model, X_cal);
    y_test_pred = predict(model, X_test);
    
    % Métricas
    r2c = calculate_r2(y_cal, y_cal_pred);
    r2p = calculate_r2(y_test, y_test_pred);
    rmsec = calculate_rmse(y_cal, y_cal_pred);
    rmsep = calculate_rmse(y_test, y_test_pred);
    
    % Criar figura
    figure('Position', [100, 100, 1500, 600]);
    
    % Gráfico Predito vs Real
    subplot(1, 2, 1);
    scatter(y_cal, y_cal_pred, 50, 'b', 'filled', 'Alpha', 0.7);
    hold on;
    scatter(y_test, y_test_pred, 50, 'r', 'filled', 'Alpha', 0.7);
    
    % Linha de referência
    min_val = min([y_cal; y_test; y_cal_pred; y_test_pred]);
    max_val = max([y_cal; y_test; y_cal_pred; y_test_pred]);
    plot([min_val, max_val], [min_val, max_val], 'k--', 'LineWidth', 2);
    
    xlabel(sprintf('Real (%s)', unit));
    ylabel(sprintf('Predito (%s)', unit));
    legend('Calibração', 'Teste', 'Location', 'best');
    title(sprintf('Predito vs Real\nR²c = %.3f, R²p = %.3f', r2c, r2p));
    grid on;
    
    % Gráfico de Resíduos
    subplot(1, 2, 2);
    residuals_cal = y_cal - y_cal_pred;
    residuals_test = y_test - y_test_pred;
    
    scatter(y_cal_pred, residuals_cal, 50, 'b', 'filled', 'Alpha', 0.7);
    hold on;
    scatter(y_test_pred, residuals_test, 50, 'r', 'filled', 'Alpha', 0.7);
    yline(0, 'k--', 'LineWidth', 2);
    
    xlabel(sprintf('Predito (%s)', unit));
    ylabel(sprintf('Resíduos (%s)', unit));
    legend('Calibração', 'Teste', 'Location', 'best');
    title(sprintf('Resíduos\nRMSEC = %.3f, RMSEP = %.3f', rmsec, rmsep));
    grid on;
    
    sgtitle('Resultados do Modelo PLS');
end

function plot_vip_scores(model, threshold)
% Plota os VIP scores
    if nargin < 2
        threshold = 1.0;
    end
    
    if isempty(model.vip_scores)
        error('VIP scores não foram calculados. Ajuste o modelo primeiro.');
    end
    
    figure('Position', [100, 100, 1200, 600]);
    
    % Plotar VIP scores
    colors = repmat({'blue'}, length(model.vip_scores), 1);
    colors(model.vip_scores > threshold) = {'red'};
    
    bar(model.vip_scores, 'FaceColor', 'flat', 'CData', ...
        [0 0 1; 1 0 0] * (model.vip_scores > threshold)' + ...
        [0 0 1; 0 0 1] * (model.vip_scores <= threshold)');
    
    % Linha de referência
    yline(threshold, 'r--', 'LineWidth', 2, ...
          'Label', sprintf('Limiar = %.1f', threshold));
    
    xlabel('Variáveis');
    ylabel('VIP Score');
    title('VIP Scores - Importância das Variáveis');
    grid on;
    
    % Rotacionar labels se necessário
    if length(model.vip_scores) > 20
        xtickangle(45);
    end
end

